### PR TITLE
bugfix(schematics): fix for angular 12 and align schematics

### PR DESCRIPTION
**Pull request by DaSchTour**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/ngneat/spectator/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #434 

## What is the new behavior?

Added the type property and also improved the behavior for angular 12 which validates the input and breaks on unkown options

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

## Other information

Sadly I couldn't test it locally because the build broke because of some rxjs issues in ActivatedRouteStub

### CODE REVIEW

File: projects/spectator/schematics/src/spectator/index.ts [hunk 1]
```diff
@@ -147,3 +147,18 @@ async function _ensurePath(tree: Tree, options: any): Promise<void> {
   options.name = parsedPath.name;
   options.path = parsedPath.path;
 }
+
+/**
+ * create a new object without given keys
+ * @param raw initial object
+ * @param keys list of keys to remove from object
+ */
+function _omit<T extends Record<string, any>>(raw: T, keys: (keyof T)[]): any {
```
Comment (benelliott) [inline projects/spectator/schematics/src/spectator/index.ts:156]: Thanks for the PR. I think this function is doing the opposite of what you want it to!

File: projects/spectator/package.json [hunk 2]
```diff
@@ -23,6 +23,7 @@
     "angular testing dumb components"
   ],
   "dependencies": {
+    "@angular-devkit/core": ">= 10.1.0",
```
Comment (NetanelBasal) [inline projects/spectator/package.json:26]: Why do we need it?

File: projects/spectator/schematics/src/spectator/index.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ import { ComponentOptions, DirectiveOptions, ServiceOptions } from './schema';
 export function spectatorComponentSchematic(options: ComponentOptions): Rule {
   return chain([
     externalSchematic('@schematics/angular', 'component', {
-      ...options,
+      ..._omit(options, ['jest', 'withHost', 'withCustomHost']),
```
Comment (NetanelBasal) [inline projects/spectator/schematics/src/spectator/index.ts:24]: Can you explain, please?

File: projects/spectator/schematics/src/spectator/index.ts (hunk 1 above)
Comment (NetanelBasal) [reply projects/spectator/schematics/src/spectator/index.ts:156]: @benelliott is right. Check your filter logic.

**Comment by NetanelBasal** [comment]
Thanks for the PR. See my comments, and please check that it works.

File: projects/spectator/package.json (hunk 2 above)
Comment (DaSchTour) [reply projects/spectator/package.json:26]: Wenn I tried to run this as a linked package it complained that @angular-devkit/core was not found although it was installed in the project.

File: projects/spectator/schematics/src/spectator/index.ts (hunk 3 above)
Comment (DaSchTour) [reply projects/spectator/schematics/src/spectator/index.ts:24]: Angular 12 complains about options that are passed but not known

**Comment by DaSchTour** [comment]
I fixed the filter logic, but I'm not able to test because build is broken.

```
Building Angular Package

------------------------------------------------------------------------------
Building entry point '@ngneat/spectator'
------------------------------------------------------------------------------
✖ Compiling TypeScript sources through NGC
ERROR: projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:44:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

44     this.params = this.paramsSubject.asObservable();
       ~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:45:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>'.

45     this.queryParams = this.queryParamsSubject.asObservable();
       ~~~~~~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:46:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Data>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Data>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

46     this.data = this.dataSubject.asObservable();
       ~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:47:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<string>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<string>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

47     this.fragment = this.fragmentSubject.asObservable() as Observable<string>;
       ~~~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:48:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").UrlSegment[]>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").UrlSegment[]>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

48     this.url = this.urlSubject.asObservable() as Observable<UrlSegment[]>;
       ~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:55:14 - error TS2416: Property 'paramMap' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>'.
    The types of 'source.operator.call' are incompatible between these types.
      Type '(subscriber: import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>, source: any) => import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/types").TeardownLogic' is not assignable to type '(subscriber: import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>, source: any) => import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/types").TeardownLogic'.
        Types of parameters 'subscriber' and 'subscriber' are incompatible.
[31[39mm          Type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>'.
            Property 'isStopped' is protected but type 'Subscriber<T>' is not a class derived from 'Subscriber<T>'.

55   public get paramMap(): Observable<ParamMap> {
                ~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:101:14 - error TS2416: Property 'root' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.
    Types of property 'paramMap' are incompatible.
      Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>'.

101   public get root(): ActivatedRouteStub {
                 ~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:105:14 - error TS2416: Property 'parent' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub | null' is not assignable to type 'ActivatedRoute | null'.
    Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.

105   public get parent(): ActivatedRouteStub | null {
                 ~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:109:14 - error TS2416: Property 'children' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub[]' is not assignable to type 'ActivatedRoute[]'.
    Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.

109   public get children(): ActivatedRouteStub[] {
                 ~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:113:14 - error TS2416: Property 'firstChild' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub | null' is not assignable to type 'ActivatedRoute | null'.
    Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.

113   public get firstChild(): ActivatedRouteStub | null {
                 ~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/create-factory.ts:105:5 - error TS2352: Conversion of type 'ActivatedRoute' to type 'SpyObject<ActivatedRouteStub>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'ActivatedRoute' is missing the following properties from type 'ActivatedRouteStub': testParams, testQueryParams, testData, testFragment, and 20 more.

105     TestBed.inject(ActivatedRoute) as SpyObject<ActivatedRouteStub>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/create-factory.ts:105:5 - error TS2352: Conversion of type 'ActivatedRoute' to type 'SpyObject<ActivatedRouteStub>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'ActivatedRoute' is not comparable to type 'ActivatedRouteStub'.

105     TestBed.inject(ActivatedRoute) as SpyObject<ActivatedRouteStub>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

An unhandled exception occurred: projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:44:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

44     this.params = this.paramsSubject.asObservable();
       ~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:45:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Params>'.

45     this.queryParams = this.queryParamsSubject.asObservable();
       ~~~~~~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:46:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Data>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").Data>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

46     this.data = this.dataSubject.asObservable();
       ~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:47:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<string>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<string>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

47     this.fragment = this.fragmentSubject.asObservable() as Observable<string>;
       ~~~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:48:5 - error TS2322: Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").UrlSegment[]>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").UrlSegment[]>'.
  Types of property 'source' are incompatible.
    Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<any>'.

48     this.url = this.urlSubject.asObservable() as Observable<UrlSegment[]>;
       ~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:55:14 - error TS2416: Property 'paramMap' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>'.
    The types of 'source.operator.call' are incompatible between these types.
      Type '(subscriber: import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>, source: any) => import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/types").TeardownLogic' is not assignable to type '(subscriber: import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>, source: any) => import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/types").TeardownLogic'.
        Types of parameters 'subscriber' and 'subscriber' are incompatible.
          Type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Subscriber").Subscriber<any>'.
            Property 'isStopped' is protected but type 'Subscriber<T>' is not a class derived from 'Subscriber<T>'.

55   public get paramMap(): Observable<ParamMap> {
                ~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:101:14 - error TS2416: Property 'root' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.
    Types of property 'paramMap' are incompatible.
      Type 'import("/Users/dasch/Projekte/spectator/projects/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>' is not assignable to type 'import("/Users/dasch/Projekte/spectator/node_modules/rxjs/internal/Observable").Observable<import("/Users/dasch/Projekte/spectator/node_modules/@angular/router/router").ParamMap>'.

101   public get root(): ActivatedRouteStub {
                 ~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:105:14 - error TS2416: Property 'parent' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub | null' is not assignable to type 'ActivatedRoute | null'.
    Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.

105   public get parent(): ActivatedRouteStub | null {
                 ~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:109:14 - error TS2416: Property 'children' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub[]' is not assignable to type 'ActivatedRoute[]'.
    Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.

109   public get children(): ActivatedRouteStub[] {
                 ~~~~~~~~
projects/spectator/src/lib/spectator-routing/activated-route-stub.ts:113:14 - error TS2416: Property 'firstChild' in type 'ActivatedRouteStub' is not assignable to the same property in base type 'ActivatedRoute'.
  Type 'ActivatedRouteStub | null' is not assignable to type 'ActivatedRoute | null'.
    Type 'ActivatedRouteStub' is not assignable to type 'ActivatedRoute'.

113   public get firstChild(): ActivatedRouteStub | null {
                 ~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/create-factory.ts:105:5 - error TS2352: Conversion of type 'ActivatedRoute' to type 'SpyObject<ActivatedRouteStub>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'ActivatedRoute' is missing the following properties from type 'ActivatedRouteStub': testParams, testQueryParams, testData, testFragment, and 20 more.

105     TestBed.inject(ActivatedRoute) as SpyObject<ActivatedRouteStub>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
projects/spectator/src/lib/spectator-routing/create-factory.ts:105:5 - error TS2352: Conversion of type 'ActivatedRoute' to type 'SpyObject<ActivatedRouteStub>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'ActivatedRoute' is not comparable to type 'ActivatedRouteStub'.

105     TestBed.inject(ActivatedRoute) as SpyObject<ActivatedRouteStub>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

**Comment by cmplank** [comment]
@DaSchTour Can you clarify about the build being broken? Are you waiting on another issue to be resolved or just stuck on this fix because your changes broke the build and you don't know how to fix it?

**Comment by DaSchTour** [comment]
@cmplank I actually have now idea what this build issue is about. For me it looks completely unrelated. But I have no idea how to fix it and why it breaks.

I just checked, the same issue is there on master. So it's not related to my changes.

**Comment by NetanelBasal** [comment]
There is a PR

**Comment by DaSchTour** [comment]
Yeah, I see. Thanks.
Closing this in favor of #457

**Pull request closed without merging**